Adding a curve3D plugin derived from Curve2D alike gnuplot's splot

cessenat opened the pull request:
### Description

<!-- Please include a summary of the change and any other information and instructions to help reviewers understand the work -->
This is a new plugin for 3D curves. It is derived from Curve2D, but when curves are separated in the file with whitespace between points, we induce the curves are segmented.

<!-- Note that all the checklist items in various bulleted lists below end with ~~ characters.
     This is to facilitate striking them out when they do not apply.
     One just has to add ~~ characters just before the opening square bracket [ -->

### Type of change

<!-- Please check one of the boxes below -->

* [ ] Bug fix~~
* [x] New feature~~
* [ ] Documentation update~~
* [ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

<!-- Please describe the tests you've added or any tests that already cover this change. Include relevant information, such as which operating system you tested on. -->

Tested on files that are provided in the PR

### Reminders:

- Please follow the [style guidelines][1] of this project.
- Please perform a self-review of your code before submitting a PR and asking others to review it.
- Please assign reviewers (see [VisIt's PR procedures][2] for more information).
I would be happy to have a review from
@biagas 

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [x] I have commented my code where applicable.
- ~~[ ] I have updated the release notes.~~
- ~~[ ] I have made corresponding changes to the documentation.~~
- [x] I have added debugging support to my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have confirmed new and existing unit tests pass locally with my changes.~~
- ~~[ ] I have added new baselines for any new tests to the repo.~~
- [x] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

cessenat committed:
Adding a curve3D plugin derived from Curve2D alike gnuplot's splot

cessenat commented:
[a3dcurve.curve3d.gz](https://github.com/visit-dav/visit/files/12821556/a3dcurve.curve3d.gz)
[a3dcurvf.curve3d.gz](https://github.com/visit-dav/visit/files/12821557/a3dcurvf.curve3d.gz)

markcmiller86 commented:
@cessenat is this a wholly *new* plugin or a swizzle on the existing (previously only 2D) curve plugins.

FWIW, I think its better (from a code maint. standpoint) to have most of the code common and shared instead of wholly separate code paths.

Any other @visit-dav/visit-developers have thoughts about this?

I have other thoughts about submitting tarfiles to be used as a PR but will wait on that until others share thoughts on the question of wholly separate plugin vs. extending existing plugins to support 3D

cessenat commented:
@markcmiller86 yes indeed this is a completely new plugin since there I use vtkUnstructuredGrid while the 2D plugin used vtkRectilinearGrid + avtCurveMetaData (which assumes 2D curves only). I think it would make the 2D plugin too complicated.
As well, I wanted to have curves segmented as gnuplot does when there is a blank line inbetween the list of points ; that is different behaviour to the one of the Curve2D plugin.

markcmiller86 commented:
> @markcmiller86 yes indeed this is a completely new plugin since there I use vtkUnstructuredGrid while the 2D plugin used vtkRectilinearGrid + avtCurveMetaData (which assumes 2D curves only). I think it would make the 2D plugin too complicated.
> As well, I wanted to have curves segmented as gnuplot does when there is a blank line inbetween the list of points ; that is different behaviour to the one of the Curve2D plugin.

Ok, thanks for that explanation. Do you have an example screenshot you can attach here showing the "...gnuplot does when there is a blank line..." That said, FWIW...I am not too swayed by these points.

cessenat commented:
![gnuplot_a3dcurve](https://github.com/visit-dav/visit/assets/40771352/dea57219-1edf-4478-8209-1ecdef41d6db)

cessenat commented:
So with VisIt now :+1: 
![visit_a3dcurve](https://github.com/visit-dav/visit/assets/40771352/1680a062-6d3f-4959-8701-d55ec2879e1e)

cessenat committed:
Update avtCurve3DWriter.C

I forgot to alter the writer for the 3D vtkUnstructuredGrid instead of the 2D vtkRectilinearGrid

cessenat committed:
Update avtCurve3DFileFormat.C

Of course in production there is no fprintf on stdout.

cessenat commented:
Maybe @markcmiller86 you could review that ?
As I said the logic in this plugin differs from that of Curve2D, so despite I got inspired by this plugin, they cannot be gathered.

markcmiller86 commented on src/databases/Curve3D/avtCurve3DFileFormat.C:
```diff
@@ -0,0 +1,1122 @@
+// Verbosity for poor man's debugging
+#define INTERACTIVEPLOT 0
+// Consider white space indicates a split
+#define MULTSEG 1
```
Would it make sense to make this a run-time `DBOptionsAttributes` as in, for example, https://github.com/visit-dav/visit/blob/7f7c4f6ed8e177b06700b412ab8a3630241341b9/src/databases/BATL/avtBATLFileFormat.C#L191-L192

cessenat replied:
Yes, good idea. As a default I will validate segmentation.
I left the MULTSEG 1 for coding check : when MULTSEG is set to 1, the option is available.

markcmiller86 commented on src/databases/Curve3D/avtCurve3DFileFormat.C:
```diff
@@ -0,0 +1,1122 @@
+//  Arguments:
+//      fname    The name of the curve file.
+//
+//  Programmer:  Hank Childs
```
Please replace all instances of other developer's name with your own and date with current.

cessenat replied:
Yes.

markcmiller86 commented on src/databases/Curve3D/avtCurve3DFileFormat.C:
```diff
@@ -0,0 +1,1122 @@
+//    Added curveTime, curveCycle.
+//
+//    Kathleen Biagas, Fri Aug 31 14:22:49 PDT 2018
+//    Added read options (currently unused).
```
In all cases, please trim out all pre-existing modification comments and just include a remark along the lines of ...

```
Copied from Curve2D plugin as of <date> and adjusted for 3D
```

cessenat replied:
OK

markcmiller86 commented on src/databases/Curve3D/avtCurve3DFileFormat.C:
```diff
@@ -0,0 +1,1122 @@
+void
+avtCurve3DFileFormat::PopulateDatabaseMetaData(avtDatabaseMetaData *md)
+{
+    if (!readFile)
```
This method is executed on both the `mdserver` and the `engine`. However, only the `engine` needs the actual curve data.

If we encounter curves with a lot of points (say > 10^6) having to *read* the *whole* file *and* store its contents in memory is bad to do on the `mdserver`. It would be better to do this...

```
if (!readFile)
{
#ifdef MDSERVER 
    bool const clearData = true;
    ReadFile(clearData);
#else
    ReadFile();
#endif  
}
```

and then down in `ReadFile(bool clearData = false)` you would clear out the `vector<double>` arrays before returning. That way, you still suffer reading all the data and briefly storing it but the `mdserver` doesn't hang onto it after its finished what it needs to populate the meta data.

cessenat replied:
Thanks for the tip.

markcmiller86 commented on src/databases/Curve3D/avtCurve3DFileFormat.C:
```diff
@@ -0,0 +1,1122 @@
+                fprintf(stdout,"\ti=%ld, no=%d, name=%s\n", i, matnos[i], names[i]);
+#endif
+            
+            mat = new avtMaterial(nmats,matnos,names,ndims,dims,0,matlist,
```
I am not sure what you are using materials for in this case? Are you using it to provide a way to break a single curve object into different pieces?

cessenat replied:
Exactly. This way a segmented curve is easy to visualize.

markcmiller86 commented on src/databases/Curve3D/avtCurve3DOptions.C:
```diff
@@ -0,0 +1,62 @@
+{
+    DBOptionsAttributes *rv = new DBOptionsAttributes;
+    vector<string> commentStyle;
+    commentStyle.push_back("# (Ultra)");
```
@biagas does Ultra support 3D curves?

biagas replied:
I don't recall, it's been a long while since I played around with Ultra.

markcmiller86 replied:
Ok, we'll leave it in in lieu of any information to the contrary.

markcmiller86 reviewed (changes requested):
@cessenat apologies for taking so long to get to reviewing this. I misunderstood the PR and thought the attached files were the plugin itself. I have made some requested changes.

In addition, we should be sure to add tests to this PR before we merge it. When you get to that point, ping me and I can provide guidance.

markcmiller86 committed:
Merge branch 'develop' into task/cessenat/2023_10_05_curve3d

cessenat committed:
Update avtCurve3DWriter.C following reviewer remarks

cessenat committed:
Update avtCurve3DWriter.h following reviewer remarks

cessenat committed:
Update avtCurve3DOptions.C following reviewer option tip

cessenat committed:
Update avtCurve3DOptions.h

cessenat committed:
Update avtCurve3DFileFormat.h following reviewer remarks

cessenat committed:
Update avtCurve3DFileFormat.C following reviewer remarks

cessenat commented:
I think i'm ready @markcmiller86

markcmiller86 commented:
@brugger1 this looks good to me. I know we don't have tests in place for it yet.

Is it ok to target this for upcoming release or will that just get in your way at this point?

brugger1 commented:
It's ok if there aren't any tests yet. You can merge it.

markcmiller86 committed:
Merge branch 'develop' into task/cessenat/2023_10_05_curve3d
